subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java

```diff
@@ -0,0 +1,503 @@
+
+                    visitor.preProperty(state);
+                    fingerprint.accept(state);
+                    visitor.postProperty();
+                }
+            }
+        });
+    }
+
+    @Nullable
+    @Override
+    public Set<String> getInputPropertiesLoadedByUnknownClassLoader() {
+        return cachingState.getInputs()
+            .map(new java.util.function.Function<CachingInputs, Set<String>>() {
+                @Nullable
+                @Override
+                public Set<String> apply(CachingInputs inputs) {
+                    ImmutableSortedSet<String> invalidInputProperties = inputs.getNonCacheableInputProperties();
+                    return !invalidInputProperties.isEmpty()
+                        ? invalidInputProperties
+                        : null;
+                }
+            })
+            .orElse(null);
+    }
+
+
+    @Override
+    public byte[] getClassLoaderHashBytes() {
+        return cachingState.getInputs()
+            .map(new java.util.function.Function<CachingInputs, byte[]>() {
+                @Nullable
+                @Override
+                public byte[] apply(CachingInputs inputs) {
+                    ImplementationSnapshot implementation = inputs.getImplementation();
+                    return implementation.getClassLoaderHash() != null
+                        ? implementation.getClassLoaderHash().toByteArray()
+                        : null;
+                }
+            })
+            .orElse(null);
+    }
+
+    @Override
+    public List<byte[]> getActionClassLoaderHashesBytes() {
+        return cachingState.getInputs()
+            .map(new java.util.function.Function<CachingInputs, List<byte[]>>() {
+                @Nullable
+                @Override
+                public List<byte[]> apply(CachingInputs inputs) {
+                    List<ImplementationSnapshot> actionImplementations = inputs.getAdditionalImplementations();
+                    if (actionImplementations.isEmpty()) {
+                        return null;
+                    }
+                    return Lists.transform(actionImplementations, new Function<ImplementationSnapshot, byte[]>() {
+                        @Override
+                        public byte[] apply(ImplementationSnapshot input) {
+                            return input.getClassLoaderHash() == null ? null : input.getClassLoaderHash().toByteArray();
+                        }
+                    });
+                }
+            })
+            .orElse(null);
+    }
+
+    @Nullable
+    @Override
+    public List<String> getActionClassNames() {
+        return cachingState.getInputs()
+            .map(new java.util.function.Function<CachingInputs, List<String>>() {
+                @Nullable
+                @Override
+                public List<String> apply(CachingInputs inputs) {
+                    List<ImplementationSnapshot> actionImplementations = inputs.getAdditionalImplementations();
```

Comment:
> I guess the variable should be called `additionalImplementations`.

You're right, I renamed the variable.



> Those are actually not the additional actions, but the actions - so they contain the task class as well. I think `getActionImplementations()` would be a perfectly good property name.
> 
> Same goes for the disabled reason: `NON_CACHEABLE_ADDITIONAL_IMPLEMENTATION` should be `NON_CACHEABLE_ACTION_IMPLEMENTATION`.

The idea here is that "additional" is additional to the implementation of the "main type", i.e. the type of the task. The task itself is only included in the additional implementations if it also declares a `@TaskAction`. I'd leave this as it is.

Reply:
I don't know why we track the task implementation at all. It doesn't seem to matter. We want to track the list of work action implementations. And I think that is enough. We don't need to track the task implementation at all, right? And for artifact transforms, we can return a list containing only the implementation of the transform action, right?

Reply:
That could work. Whether it's a good idea or not depends on what type of usage we optimize for.

Right now we have a primary type, and for tasks we keep track of some additional stuff on the side. In most cases the additional stuff should be empty (right now this is modeled by the list containing the primary type as a single element). We actually log a message if the task has additional actions, because it's usually a sign of problems with cacheable tasks.

In the future I think the goal will be to have each implementation associated with its own inputs and outputs, hence "additional" implementations will be retired. You'll still be able to add actions before and after a task, but they will be individual work items with their own inputs and outputs.

There's a relatively clear path to this future, and the road ends with "additional implementations" disappearing entirely from even build scans.

If we change this, and make the primary implementation plural, that gives us a lot of work on both BT and GE side, and when we reach the state of single implementation, we need to change things again.

I think staying with the current setup is best.